eppo_client/bandit.py

```diff
@@ -0,0 +1,267 @@
+            flag_key, subject_key, action_weights
+        )
+        return BanditEvaluation(
+            flag_key,
+            subject_key,
+            subject_attributes,
+            selected_action,
+            actions_with_contexts[selected_idx].attributes,
+            action_scores[selected_idx][1],
+            action_weights[selected_idx][1],
+            bandit_model.gamma,
+        )
+
+    def score_actions(
+        self,
+        subject_attributes: Attributes,
+        actions_with_contexts: List[ActionContext],
+        bandit_model: BanditModelData,
+    ) -> List[Tuple[str, float]]:
+        return [
+            (
+                action_context.action_key,
+                (
+                    score_action(
+                        subject_attributes,
+                        action_context.attributes,
+                        bandit_model.coefficients[action_context.action_key],
+                    )
+                    if action_context.action_key in bandit_model.coefficients
+                    else bandit_model.default_action_score
+                ),
+            )
+            for action_context in actions_with_contexts
+        ]
+
+    def weigh_actions(
+        self, action_scores, gamma, probability_floor
+    ) -> List[Tuple[str, float]]:
+        number_of_actions = len(action_scores)
+        best_action, best_score = max(action_scores, key=lambda t: t[1])
+
+        # adjust probability floor for number of actions to control the sum
+        min_probability = probability_floor / number_of_actions
+
+        # weight all but the best action
+        weights = [
+            (
+                action_key,
+                max(
+                    min_probability,
+                    1.0 / (number_of_actions + gamma * (best_score - score)),
+                ),
+            )
+            for action_key, score in action_scores
+            if action_key != best_action
+        ]
+
+        # remaining weight goes to best action
+        remaining_weight = 1.0 - sum(weight for _, weight in weights)
+        weights.append((best_action, remaining_weight))
+        return weights
+
+    def select_action(self, flag_key, subject_key, action_weights) -> Tuple[int, str]:
+        # deterministic ordering
+        sorted_action_weights = sorted(
+            action_weights,
+            key=lambda t: self.sharder.get_shard(
+                f"{flag_key}-{subject_key}-{t[0]}", self.total_shards
+            ),
+        )
+
+        # select action based on weights
+        shard = self.sharder.get_shard(f"{flag_key}-{subject_key}", self.total_shards)
+        cumulative_weight = 0.0
+        shard_value = shard / self.total_shards
+
+        for idx, (action_key, weight) in enumerate(sorted_action_weights):
+            cumulative_weight += weight
+            if cumulative_weight > shard_value:
+                return idx, action_key
+
+        # If no action is selected, return the last action (fallback)
```

Comment:
This would be very unexpected given the top scoring action should fill the rest of the space--thinking we throw an error instead ([src](https://github.com/Eppo-exp/java-server-sdk/blob/main/src/main/java/com/eppo/sdk/helpers/VariationHelper.java#L19))

Reply:
I don't want to throw an error at runtime in the SDK -- let's log an error though

Reply:
Yeah I was assuming there would be some error handler at the top-level (unless graceful mode is off) that catches and logs an error and returns the default value.

Otherwise we could get in a situation where a bug is introduced and all bandits start selecting the last action all the time--seems worse than returning the default value

Reply:
You've convinced me, also did not add top level error handling -- fixed